### PR TITLE
Add Complement Objective Training

### PR DESCRIPTION
**Pull request by yashkant**
Motivation
---
In a recently [[accepted work](https://openreview.net/forum?id=HyM7AiA5YX)] at ICLR 2019 a new training scheme called Complement Objective Training (COT) was proposed. Empirical results have shown that COT is capable of improving the performance of all the state-of-the art classification models. This PR adds COT to Pythia. 

In the original work the authors only focus on multi-class (one-of-many) classification problems. Since Pythia is a multi-label classification model, I have made some changes to COT to suit for this task. 

This [[report](https://drive.google.com/file/d/16NtLvZvBPq1cRVeCSq7sXXg0C8NkSi4l/view)] summarizes the approach and modifications done (on top of the original work) for adding it to Pythia.

Overall, I believe that COT could possibly improve on the earlier performance of Pythia. 

Summary of Code Changes 
---
Changes have been made to the following files:

- **`train_model/Loss.py`**: 
	- `get_loss_criterion()` function has been updated and now handles a list of losses (with maximum length 2). Currently only supports using `complementLoss` with `softmaxKL` as primary loss (please refer to Section 2.5 of the [[report](https://drive.google.com/file/d/16NtLvZvBPq1cRVeCSq7sXXg0C8NkSi4l/view)] for more justification on this). 
	-  `ComplementEntropyLoss` class and `complement_entropy_loss()` helper function  are added.
	- `CombinedLoss` class now supports mixing `complementEntropy` loss alongwith the already existing  `softmaxKL` and `logitBCE` losses.

- **`train.py`**: 
	- Now builds  a list of `losses`, `optimizers`, `schedulers` and passes them to the `one_stage_train()` function.
	- Each loss has a separate optimizer and scheduler. 

- **`config/config.py`**: 
	- New configuration options `use_complement_loss (boolean)` and `complement_loss (string)`   are added. 

- **`train_model/Engineer.py`**:  

	- `save_a_report()` and  `save_a_snapshot()` functions now handle `train_losses`,  `loss_criterions` and `optimizer_list` parameters as lists.
	- Some refactoring has been done to preserve 80 character line length (previously violated). 
	- `one_stage_train()` function now handles `train_losses`, `optimizer_list` `loss_criterions` as lists, training loop has been modified to support the alternate training scheme.
	- `evaluate_a_batch()` and `compute_a_batch()` functions can now handle losses either passed as `nn.Module` object or list of objects. They return a list of tensors for a list input or tensor for an object input respectively. **(edited for clarity)**
	- `one_stage_eval()` can handle list of `loss_criterions` or single object and returns correspondingly. 
	
I do not find any changes that will affect the outer APIs hence the `README.md` was not modified.

Results
---
Due to the unresolved issue #30  (without the extracted image features, detectron model) I could not report on the new results from COT. However, I have tested the `complementEntropy` loss  and the `one_stage_train()` function in the modified setup (using dummy data) and verified that this piece of code works.

I plan to run the experiments and add the results as soon as issue #30 is resolved. Please let me know if any changes/clarifications are needed. 

I have signed the CLA.

### CODE REVIEW

**Comment by yashkant** [comment]
Experiments
---
I have finished running experiments with Complement Objective Training. The [[report](https://drive.google.com/file/d/1glomNqP1TZNA7iub1RLUZPnMFfx8-XGf/view)] has been updated with the results, observations and discussion. 

To set the reference, I have run Pythia’s baseline experiment using the default configuration same as [[config.py](https://github.com/facebookresearch/pythia/blob/master/config/config.py)] with Softmax KL Divergence and Binary Cross Entropy losses. All the experiments ahead are performed by making changes on top of this default configuration. 

Next, I have run the Softmax KL Divergence loss baseline experiment by coupling it with Complement Objective Training. Training with COT fails to attain convergence and the final performance keeps fluctuating around 30%.  Exact results obtained from the experiments are shown below:

| Loss| mini-val (%)|test-dev(%)|
|---|---| ---|
| Softmax KL Divergence| 65.049|66.86|
| Binary Cross Entropy|**66.158**|**67.75**|
| Softmax KL Divergence with COT| 29.19 (no convergence)|31.6 (no convergence)|


From observations, it could be claimed (to a good degree of certainty) that complement training was conflicting with the primary training. This is discussed in Section 3.2 an d 3.4 of the [[report](https://drive.google.com/file/d/1glomNqP1TZNA7iub1RLUZPnMFfx8-XGf/view)].

The convergence issue was resolved by mixing the complement loss and primary loss in a single combined loss formulation using a hyperparameter weight . For more discussion on why this works this please refer to Section 3.4 of the [[report](https://drive.google.com/file/d/1glomNqP1TZNA7iub1RLUZPnMFfx8-XGf/view)].

Let   denote the primary loss, complement loss respectively and let  be a hyperparameter weight. The combined loss $K(\mathbf{y}, \mathbf{\hat{y}})$ can be then formulated as:



Results obtained by using the combined loss are shown below:  

| Loss|  |mini-val (%)|test-dev(%)|
|---|---|---| ---|
| Softmax KL Divergence + Complement| 0.1|65.068|66.86|
|  Softmax KL Divergence + Complement| 0.5|**65.096**|66.74|
|  Softmax KL Divergence + Complement| 1.0|**65.096**|**66.95**|
|  Softmax KL Divergence + Complement| 10.0|64.880|66.72|



Experiments were also performed by using a decay schedule on $\alpha$. It was decayed at discrete steps with a fixed decay factor during the training cycle (refer to Section 3.3 in [[report](https://drive.google.com/file/d/1glomNqP1TZNA7iub1RLUZPnMFfx8-XGf/view)]). The results obtained are shown in the table below:

| Loss| initial    value|  decay factor|  decay iters |mini-val (%)|test-dev(%)|
|---|---| ---|---|---|---|
|  SKLD + Complement| 10|0.1|1000|65.186|66.83
|  SKLD + Complement| 25|0.5|1500|**65.193**|66.77
|  SKLD + Complement| 100|0.01|3000|65.101|**66.84**
|  SKLD + Complement| 128|0.5|500|65.126|66.82
|  SKLD + Complement| 1000|0.05|500|65.177|66.76
  

Overall, I do not find any significant improvement in the performance of Pythia after adding the Complement Objective Training.
  
#### Result files for the above experiments are uploaded here: [[Google Drive Link](https://drive.google.com/drive/folders/1G6MYds5LyQOjG8h-pchEhCPtXfAAKbHa)]

Running the Experiments
---
 To run with Complement Training experiment use: 

	python train.py --config_overwrite '{use_complement_loss:true, loss:"softmaxKL", complement_loss:"complementEntropy"}'


 To run the default Combined Loss experiment use: 

	python train.py --config_overwrite '{loss:"combinedLoss"}'

To run the Combined Loss with Weight Decay experiment use: 

	python train.py --config_overwrite '{loss:"combinedLoss", weight_complement:10.0, weight_complement_decay:true, weight_complement_decay_factor:0.1, weight_complement_decay_iters:2000}'


More configurations that are placed in [[config.py](https://github.com/facebookresearch/pythia/blob/master/config/config.py)] can be used similarly.


Summary of Additional Code Changes
--- 
The codebase has been migrated to Pytorch 1.0 following the migration guide.

- **`config/config.py`**: 
	- New configuration options added for combined loss training with weight decay scheme (please refer Section 3.3 in the [[report](https://drive.google.com/file/d/1glomNqP1TZNA7iub1RLUZPnMFfx8-XGf/view)]).
	- New configurations are added for separate gradient clipping on complement loss.  

- **`top_down_bottom_up/image_attention.py`**: 
	- Added a small constant during the division operation of the `masked_attention` layer to fix Nan gradients issue while backpropogating with complement loss. 

- **`train_model/Loss.py`**: 
	- Loss function APIs are modified in order to accept a new *optional* argument `iter` that helps in decay of weight hyperparameter .  
	- `CombinedLoss` class has been rewritten to include mixing `ComplementEntropy` loss. The corresponding `weight` and `decay routine` parameters are added in `config/config.py` file. 

- **`train_model/Engineer.py`**: 
	- `clip_gradients()` function now accepts an additional argument `grads_label`  which helps to identify gradients from primary and complement losses. 
	- Loss function calls now pass `iter` as an argument that helps in decay schedule of hyperparameter weight . 

- **`README.md`**: 
	- Now installs Pytorch version 1.0.1 instead of version 0.3.1.


Additional Experiments
---
Code for the experiments mentioned in appendix of the [[report](https://drive.google.com/file/d/1glomNqP1TZNA7iub1RLUZPnMFfx8-XGf/view)] lie in a separate branch  [[add-cot-test]](https://github.com/yashkant/pythia/tree/add-cot-test) in my fork. It includes:  

- Experiment to cross-verify the COT implementation (used in Pythia) against the author's implementation, this was done in order to make sure that COT implementation had no bugs. The code is kept inside `verify_cot/` and directions to run the experiment are [[here](https://github.com/yashkant/pythia/blob/add-cot-test/verify_cot/COT/README.md)].  Please refer to `Appendix A.1` for more information on this. 

 - Configurations to run experiments that are mentioned in `Appendix A.2` of the [[report](https://drive.google.com/file/d/1glomNqP1TZNA7iub1RLUZPnMFfx8-XGf/view)]. 

 - Code to visualise gradient flow across the model (used for debugging). Use `show_gradients` config in `config/config.py` for this.

Result files for the experiments in `Appendix A.2` are also uploaded to the [[Google Drive Link](https://drive.google.com/drive/folders/1G6MYds5LyQOjG8h-pchEhCPtXfAAKbHa)] and are prefixed with `Add-{ExpName}`. 


Conclusion
---
Overall, using Complement Objective Training does not show any substantial improvement in Pythia's performance and seems undesirable in its present state. 

If any of the features from this PR are desirable in Pythia (migration to Pytorch 1.0, etc), I would be happy to make a separate PR for it.

**Comment by apsdehal** [comment]
@yashkant Sorry for really late reply. We had an internal version, ready to made public, so we had to delegate this a bit. If you can adjust this PR according to new version then it would be really helpful.

**Comment by yashkant** [comment]
Hi @apsdehal,

I did not find any significant improvements after coupling COT with the Pythia's version 0.1, largely because the complement objective that was proposed in the original work did not behave as intended with Pythia. 

I will adjust this PR accordingly for the newer version, however, I do not anticipate very significant improvements from this. 

Also, please expect a little delay.

**Comment by apsdehal** [comment]
Closing as too old. Please open up a new PR if you plan to contribute. :)
